### PR TITLE
Refactor & pybind11

### PR DESCRIPTION
**Pull request by breznak**
This PR
- refactors external libs into `/include/` and (optional) `external/` (where sources are)

So far: 
 - [x] boost moved (/include, /external/boost)
  - and updated to 1.66
  - added `/include` to CMake

Reference: #31

### CODE REVIEW

File: external/boost/INSTALL [hunk 1]
```diff
@@ -0,0 +1,8 @@
+See ./index.html for information about this release. The "Getting Started" 
+section is a useful starting place.
```
Comment (breznak) [inline external/boost/INSTALL]: do we need the sources /external/boost? We don't build it automatically, and it mostly duplicates the headers in /include

File: CMakeLists.txt [hunk 2]
```diff
@@ -81,6 +81,7 @@ include(CommonCompilerConfig)
 # (see individual external/*.cmake modules for exported settings and functions)
 add_subdirectory(external)
 include_directories(SYSTEM ${EXTERNAL_INCLUDE_DIRS})
+include_directories(SYSTEM ${REPOSITORY_DIR}/include/) #all headers reside here (boost,...)
```
Comment (breznak) [inline CMakeLists.txt]: add the new header path to CMake

File: external/boost/INSTALL (hunk 1 above)
Comment (chhenning) [reply external/boost/INSTALL]: we need to build at least two libs. boost.filesystem and boost.system.

Comment (breznak) [reply external/boost/INSTALL]: but we are not currently building it. And boost doesn't have CMake, which makes it complicated. Do you use the bat/sh scripts to build boost? Also, how to limit it only to those sub-libs of our interest, so the build doesn't take ages?

Comment (chhenning) [reply external/boost/INSTALL]: Have a look here:
http://www.boost.org/doc/libs/1_66_0/more/getting_started/unix-variants.html
http://www.boost.org/doc/libs/1_66_0/more/getting_started/windows.html

In essence:
1. you need to build the build system with the bootstrap script
2. run this command line:
b2 --toolset=msvc variant=release link=static runtime-link=shared address-model=64 --with-system --with-filesystem stage

please not that toolset will change depending on the platform

Comment (mattheiler) [reply external/boost/INSTALL]: For CMake, what do you think of [this](https://discourse.numenta.org/t/community-fork-cmake-enhancements-and-package-management/3370/8) suggestion?

We'd basically create a repository for package management - building all of our dependencies and packaging them for different target triplets.

The CMake would then download that package (e.g., x64-linux-gcc.tar.gz), unzip it, then throw it in the bin, include, etc.

```
/boost
  /1.66
    x64-linux-gcc.tar.gz
    x64-windows-gcc.zip
    x64-windows-msvc.zip
    x86-linux-gcc.tar.gz
    x86-windows-gcc.zip
    x86-windows-msvc.zip
```

If someone needs a new package (e.g. boost@1.7:x64-windows-gcc), they'd build all the dependencies using that repository. PRs would be for new packages. We could deprecate old packages we don't want to support.

I think this gives us a lot of control over our dependencies; building them in a separate process, but quickly pulling them into different environments and making sure they're the same binaries for everyone. We could use apt-get and potentially vcpkg to get them and package them, although see the issue in the forum topic for issues with vcpkg.

Comment (chhenning) [reply external/boost/INSTALL]: For now I'm fine with whatever works. :-)

Comment (mattheiler) [reply external/boost/INSTALL]: Okay, do what @chhenning did. I'm proposing long-term. Maybe we should open a separate issue for package management.

Comment (chhenning) [reply external/boost/INSTALL]: For short term can we provide binaries? At least for Visual Studio 2017 that's what I'm doing.

Comment (mattheiler) [reply external/boost/INSTALL]: Yeah, I'm even proposing we do that in the long run. I'd just organize them by target triplet (i.e., [machine]/[vendor]/[operating system]). We should just have a common way of producing those binaries for different triplets.

Comment (breznak) [reply external/boost/INSTALL]: @chhenning where do you use the boost lib files (filesystem, dll)? Currently we managed to do only with headers. 

@mattheiler the repo idea: we had that, as nupic-linux64, nupic-osx64, ... basically bootstrap repos with all the deps preinstalled. For some reason we abandoned that idea...would be prolly a back-to-forum topic. 

or how difficult is it for the 3 platforms to install boost? We just say download it here and install?

Comment (chhenning) [reply external/boost/INSTALL]: We need boost.filesystem and boost.dll in the os layer. Some code has been commented out there for now.
On Windows, I need to link with boost.filesystem and boost.system right now.

**Comment by breznak** [comment]
Will be replaced with another PR

**Pull request closed without merging**